### PR TITLE
Add `HtmlTag#withSelf`, restore debounce example

### PR DESCRIPTION
**Pull request by armanbilge**
Closes https://github.com/armanbilge/calico/issues/132.

### CODE REVIEW

File: docs/time.md [hunk 1]
```diff
@@ -89,7 +89,7 @@ val app = Channel.unbounded[IO, String].toResource.flatMap { emailCh =>
     div(
       span(
         label("Your email: "),
-        input { self =>
+        input { (self: fs2.dom.HtmlInputElement[IO]) => // FIXME compiler bug
           onInput --> (_.evalMap(_ => self.value.get).through(emailCh.sendAll))
```
Comment (armanbilge) [inline docs/time.md:93]: This is extremely annoying. Another workaround is to do:
```scala
input { self =>
  Tuple1(onInput --> (_.evalMap(_ => self.value.get).through(emailCh.sendAll)))
}
```

I'm going to try and minimize and report a compiler bug.

Comment (armanbilge) [reply docs/time.md:93]: - https://github.com/lampepfl/dotty/issues/16737

**Review by 2chilled — DISMISSED** [review]
Seems like there's still work to do in regards to implementing overloads soundly, even in Scala 3. To this day, I've always avoided them without missing too much.

**Comment by armanbilge** [comment]
> To this day, I've always avoided them without missing too much.

Yes, that is often good practice. Do you think we should do that here? We can add it as an alternative API e.g.

```scala
input.withSelf { self =>
  onInput --> (_.evalMap(_ => self.value.get).through(emailCh.sendAll))
}
```

I think Laminar might call it `withContext`.

**Comment by 2chilled** [comment]
I'd absolutely prefer that, yes. These types of bugs are super annoying, especially for non expert Scala users. In my experience it's best to even disable all nasty features that can be somehow disabled. At work, we use the strictest scalac 2 settings together with nearly all of https://www.wartremover.org/, and everyone is super happy with that. Furthermore, we still avoid Scala 3 there, because there are a bunch of strict flags not implemented (yet?).

**Comment by 2chilled** [comment]
Ah and `withSelf` is a good name in my subjective eyes :)

**Comment by armanbilge** [comment]
Great, I'll add that!

> because there are a bunch of strict flags not implemented (yet?).

Good news, many new strict flags are arriving in Scala 3.3.0 :)

- https://github.com/lampepfl/dotty/pull/16157

File: calico/src/main/scala/calico/html/HtmlTag.scala [hunk 2]
```diff
@@ -31,7 +31,7 @@ final class HtmlTag[F[_], E] private[calico] (name: String, void: Boolean)(using
   def apply[M](modifier: M)(using M: Modifier[F, E, M]): Resource[F, E] =
     build.toResource.flatTap(M.modify(modifier, _))
 
-  def apply[M](mkModifier: E => M)(using M: Modifier[F, E, M]): Resource[F, E] =
+  def withSelf[M](mkModifier: E => M)(using M: Modifier[F, E, M]): Resource[F, E] =
```
Comment (armanbilge) [inline calico/src/main/scala/calico/html/HtmlTag.scala:34]: Not to second guess 😅 maybe another option is `fromSelf`. But `withSelf` is good. I think 😅

File: docs/time.md [hunk 3]
```diff
@@ -89,7 +89,7 @@ val app = Channel.unbounded[IO, String].toResource.flatMap { emailCh =>
     div(
       span(
         label("Your email: "),
-        input { (self: fs2.dom.HtmlInputElement[IO]) => // FIXME compiler bug
+        input.withSelf { self =>
```
Comment (armanbilge) [inline docs/time.md:92]: Yeah, overloads are bad 😓

**Review by 2chilled — APPROVED** [review]

**Comment by armanbilge** [comment]
> Seems like there's still work to do in regards to implementing overloads soundly, even in Scala 3.

Yeah, seems like what I was trying to do here was an unsupported behavior according to https://github.com/lampepfl/dotty/issues/16737#issuecomment-1400531858 😅

**Pull request merged**